Make select forms look more like other forms.
Insert "[_" and "_]" as part of the form link, and
include it as part of the link.
By: yozohida@gmail.com
---
 src/document/html/parser/forms.c |    4 +-
 src/viewer/text/form.c           |   42 ++++++++++++++++++++++++++++----------
 src/viewer/text/link.c           |    2 +-
 3 files changed, 34 insertions(+), 14 deletions(-)

Index: elinks/src/document/html/parser/forms.c
===================================================================
--- elinks.orig/src/document/html/parser/forms.c	2017-11-01 03:05:25.036072431 +0100
+++ elinks/src/document/html/parser/forms.c	2017-11-01 03:05:25.036072431 +0100
@@ -494,10 +494,10 @@
 	fc->labels = labels;
 
 	menu_labels(fc->menu, "", labels);
-	put_chrs(html_context, "[", 1);
 	html_stack_dup(html_context, ELEMENT_KILLABLE);
 	format.form = fc;
 	format.style.attr |= AT_BOLD;
+	put_chrs(html_context, "[&nbsp;", 7);
 
 	max_width = 0;
 	for (i = 0; i < order; i++) {
@@ -514,8 +514,8 @@
 	for (i = 0; i < max_width; i++)
 		put_chrs(html_context, "_", 1);
 
+	put_chrs(html_context, "&nbsp;]", 7);
 	pop_html_element(html_context);
-	put_chrs(html_context, "]", 1);
 	html_context->special_f(html_context, SP_CONTROL, fc);
 }
 
Index: elinks/src/viewer/text/form.c
===================================================================
--- elinks.orig/src/viewer/text/form.c	2017-11-01 03:05:25.036072431 +0100
+++ elinks/src/viewer/text/form.c	2017-11-01 03:12:34.273791002 +0100
@@ -589,28 +589,38 @@
 			else
 				/* XXX: when can this happen? --pasky */
 				s = "";
-#ifdef CONFIG_UTF8
-			if (term->utf8_cp) goto utf8_select;
-#endif /* CONFIG_UTF8 */
+
+			/* insert preceding '[_' */
+			i = 0;
+			x = link->points[i].x + dx;
+			y = link->points[i].y + dy;
+			if (is_in_box(box, x, y))
+				draw_char_data(term, x, y, '[');
+			i++;
+			x = link->points[i].x + dx;
+			y = link->points[i].y + dy;
+			if (is_in_box(box, x, y))
+				draw_char_data(term, x, y, '_');
+			i++;
+
+#ifndef CONFIG_UTF8
 			len = s ? strlen(s) : 0;
-			for (i = 0; i < link->npoints; i++) {
+			for (; i < link->npoints-2; i++) {
 				x = link->points[i].x + dx;
 				y = link->points[i].y + dy;
 				if (is_in_box(box, x, y))
-					draw_char_data(term, x, y, i < len ? s[i] : '_');
+					draw_char_data(term, x, y, i-2 < len ? s[i] : '_');
 			}
-			break;
-#ifdef CONFIG_UTF8
-utf8_select:
+#else
 			text = s;
 			end = strchr((const char *)s, '\0');
 			len = utf8_ptr2cells(text, end);
-			for (i = 0; i < link->npoints; i++) {
+			for (; i < link->npoints-2; i++) {
 				x = link->points[i].x + dx;
 				y = link->points[i].y + dy;
 				if (is_in_box(box, x, y)) {
 					unicode_val_T data;
-					if (i < len) {
+					if (i-2 < len) {
 						int cell;
 
 						data = utf8_to_unicode(&s, end);
@@ -628,8 +638,18 @@
 					draw_char_data(term, x, y, data);
 				}
 			}
-			break;
 #endif /* CONFIG_UTF8 */
+			/* insert trailing ' ]'. */
+			x = link->points[i].x + dx;
+			y = link->points[i].y + dy;
+			if (is_in_box(box, x, y))
+				draw_char_data(term, x, y, '_');
+			i++;
+			x = link->points[i].x + dx;
+			y = link->points[i].y + dy;
+			if (is_in_box(box, x, y))
+				draw_char_data(term, x, y, ']');
+			break;
 		case FC_SUBMIT:
 		case FC_IMAGE:
 		case FC_RESET:
Index: elinks/src/viewer/text/link.c
===================================================================
--- elinks.orig/src/viewer/text/link.c	2017-11-01 03:05:25.036072431 +0100
+++ elinks/src/viewer/text/link.c	2017-11-01 03:05:25.036072431 +0100
@@ -123,6 +123,7 @@
 		case LINK_CHECKBOX:
 			return 1;
 
+		case LINK_SELECT:
 		case LINK_BUTTON:
 			return 2;
 
@@ -156,7 +157,6 @@
 
 		case LINK_HYPERTEXT:
 		case LINK_MAP:
-		case LINK_SELECT:
 			return 0;
 	}
 
